### DISCUSSION TITLE
Best practice regarding outliers for dynamic topic modeling

### DISCUSSION THREAD

**Discussion by serenalotreck** [MaartenGr/BERTopic#1738] [Q&A] [answered]
Hi!

I'm working on dynamic topic modeling for a corpus of scientific abstracts, and I just wanted to check in regarding a best practice that I didn't see mentioned in the docs.

The docs do talk about outlier reduction, which forces all points to be assigned to a cluster. I performed this on my data, as of my original 5,564 abstracts, 2,694 were outliers when I first ran the model.

I then performed topic modeling with:
```
topics_over_time = topic_model.topics_over_time(docs, abstracts['year'], global_tuning=False, nr_bins=50)
```

I had the thought that maybe the topics over time would be clearer if I didn't reduce outliers, because (to my understanding of how this works) they might be weakening the specific signal of the topic clusters since there were so many of them (I'm not sure if that's true).

I did notice quite a difference between the two. Here's the plot of the results with outlier reduction:

![image](https://github.com/MaartenGr/BERTopic/assets/41377532/d60147ad-d6fa-40b3-b619-78a576f484fe)

And without outlier reduction:

![image](https://github.com/MaartenGr/BERTopic/assets/41377532/69a43a67-d784-4062-83ba-351fbaebec48)

The difference in frequencies is large enough that it would change my interpretation of the popularity of topics relative to one another over time.

I was wondering if folks had thoughts on which plot is the "truer" representation of topic evolution over time?

EDIT: Realized I posted the wrong screenshots! The previous ones were from two separate models with different random seeds; now that I'm looking at both version from the same model, it looks like the impact isn't that large. I'd imagine it could be, so would still love to know people's thoughts!

**Comment by MaartenGr** [comment, marked as answer]
HDBSCAN can be quite strict in assigning outliers. Therefore, `reduce_outliers` was introduced to reduce its impact. In practice, you would have to check yourself whether the new assignments make sense by inspecting a subset manually. Do note that this function can also reduce some outliers and it's actually advised not to reduce all of them.

**Reply by lintonye** [reply]
@MaartenGr I ran `fit_transform` again on the outliers only, which seemed to produce some cohesive topics (and therefore reducing outliers), and then merged this new model with the original one. This seems to work better for my use case since `reduce_outliers` does not create new topics and in a sense undermine the original topics.

Does this sound like a reasonable approach? 

BTW: I couldn't find a way to remove topics or docs from a model - wanted to remove the outliers from the original model. I'd appreciate any pointers.

**Reply by MaartenGr** [reply]
@lintonye It definitely is! There are a bunch of methods/techniques you could use in order to reduce outliers but the one you describe is a solid one. Do note though that you are likely to end up with more micro-clusters than you initially had. 

Removing topics from a model is currently not supported as during inference, you could simply use `calculate_probabilities=True` and ignore the outlier topic. Also, if it is a non-outlier topic that you want to remove, you can simply merge it with the outlier topic.

**Reply by lintonye** [reply]
@MaartenGr Thanks! This seems to work well so far. Since removing topics/docs isn't supported yet, I had to store the initial documents with some duplication so that I could call `get_document_info` at a later time. Why does `get_document_info` always require the initial list of documents?

Does it make sense to mention this approach in the documentation? I could prepare a PR if it is.

**Reply by lintonye** [reply]
Also, when merging two models, will some non-outlier documents be assigned as outliers? I've noticed size discrepancy after merging. For example, let's say initially the docs are:

Model 1
```
doc1, topic: -1
doc2, topic: -1
doc3, topic: -1
doc4, topic: 0
doc5, topic: 1
```

To reduce outliers, we retrain a model with `doc1`, `doc2` and `doc3` and get:

Model 2
```
doc1, topic: -1
doc2, topic: 0
doc3, topic: 1
```

Now if I merge these two models, I expect to get something like:

```
doc1, topic: -1
doc2, topic: -1
doc3, topic: -1
doc1, topic: -1 ==> duplicate due to merge
doc4, topic: 0
doc5, topic: 1
doc2, topic: 0
doc3, topic: 1
```

So, the number of non-outliers is 4, which is [total # of docs] - [# of outliers in model 2].

However, in my tests, the number of non-outliers could be 3, which means for example doc2 gets assigned as an outlier in the merged model. Is this expected?

**Reply by MaartenGr** [reply]
It might happen if the minimum similarity is quite low or when a given topic might just be very similar to the outliers (although I wouldn't expect the latter to happen frequently).

**Reply by bfisseler** [reply]
Dear @lintonye, dear @MaartenGr,

I am relatively new to BERTopic and ran into the problem of having (too) many outliers. Taking these outliers and fit a new model on them seems like a good approach. Do you mind providing an example of how to achieve this? I tried to simply extract the outliers, fit a new model and then merge_models(). Then I tried to remove the outliers from the original model before merging. But both approaches produce a colorful dot pattern when I run visualize_documents() instead of the relatively clear clusters from the original model. What am I missing?

**Reply by MaartenGr** [reply]
@bfisseler Hmmm, that's difficult to say. Could you perhaps create a new issue and provide an example? There might be a bug or perhaps something that can be improved with your code.

**Comment by ioannamkk** [comment]
Hi, I am new to BERTopic and am working on an analysis of news articles about a specific climate policy (n = 6,131).

I have the following setup:

- Embedding model: all-MiniLM-L6-v2 with chunking (articles are long, median ~925 tokens)
- UMAP: n_neighbors=10, n_components=15, min_dist=0.0
- HDBSCAN: min_cluster_size=30, min_samples=5, cluster_selection_method='eom'

My model produces 54 topics with ~25% outliers. I have read on this forum that it is advised not to remove all outliers, as some documents are legitimately uncategorizable. My question is:
- For a news corpus covering a political and policy debate, is 25% a reasonable outlier rate to report and justify, or should I be trying to reduce it further?
- I tried to lower min_samples from 5 to 3, which had almost no effect. I also tested n_neighbors=5 and min_cluster_size=10, which produced 161 topics but still ~21% outliers.

Any guidance would be really appreciated. Thank you!

**Reply by MaartenGr** [reply]
> For a news corpus covering a political and policy debate, is 25% a reasonable outlier rate to report and justify, or should I be trying to reduce it further?

Do you think it is reasonable? What is and isn't reasonable really depends on the dataset itself. The definition of an outlier is in algorithms a mathematical one and might not align with domain expertise. As such, it's important to look at the data yourself and see whether it should be reduced. If so, then I would advise looking through the FAQ.
